python/dune/xt/common/__init__.py

```diff
@@ -12,13 +12,33 @@
 
 from importlib import import_module
 
+# This is here on purpose, needs to be imported as early as possible!
 try:
     from mpi4py import MPI
 except ImportError:
     pass
 
-import dune.xt
-from dune.xt._common import *
-init_logger = dune.xt._common._init_logger
-test_logger = dune.xt._common._test_logger
-init_mpi = dune.xt._common._init_mpi
\ No newline at end of file
+# empty is missing here on purpose!
+_modules = (
+    '_common',
+    '_logging',
+    '_timings',
+    )
+
+# see https://stackoverflow.com/questions/43059267/how-to-do-from-module-import-using-importlib
+for mod_name in _modules:
+    try:
+        mod = import_module('.{}'.format(mod_name), 'dune.xt.common')
+        if "__all__" in mod.__dict__:
+            names = mod.__dict__["__all__"]
+        else:
+            # otherwise we import all names that don't begin with _
+            names = [x for x in mod.__dict__ if not x.startswith("_")]
+        globals().update({k: getattr(mod, k) for k in names})
```

Comment:
This unconditionally overwrites existing objects in globals. I think it would be better if that was guarded. 
It's also functionally different from before, no? 
I'd be happier with how we've done this in the logging module atm: https://github.com/dune-community/dune-xt-common/blob/6eab6f6a30f25436266b1b75ac4bd7177cbc2291/python/dune/xt/common/logging.py

Reply:
So, what I wanted to achieve is the following:

* guard the import of stuff from `.so`s and give a warning if that fails
* do this for lots of `.so`s without code duplication
* have deterministic ordering of the modules in one place (so we do not have tom import dependent modules within the `.so`, which is hard to debug and leads to lots of imports).

See for instance https://github.com/dune-community/dune-xt-functions/blob/dailywork-ftschindler/python/dune/xt/functions/__init__.py#L20 (there will probably be even more modules in the future), where this all originated from.

I could just do it as in the `logging` example, move this to the `__init__.py` and additionally guard the import, but then I would have to write the code for each of the `.so`s, which I do not want to do. Or am I missing something there?

Reply:
Considering guarding the imports I think we should note what problem this actually solves. In common this was due to the fact that the cmake.py code needs to be executable at configure time when no bindings could've been built yet. I do not see the need for it in other places atm.

I feel like the mandatory ordering in the importing module is a broken design. The library has no way to actually enforce this. The user can still directly import symbols from the .so. I'm not sure I understand what the issue with debugging imports in the .so is. 
I think if every module did its imports correctly w/o the 'import module ordering' there would be much less chance of problems. We can still add the global reg for types if that duplication of wrapped types is the issue you're thinking of.

Reply:
> This unconditionally overwrites existing objects in globals. I think it would be better if that was guarded.

```python
from _foo import *
from _bar import *
```
also overwrites everything, and we've been doing that in some places.
I mainly want to get an error during import which `.so` is broken, so I know where to start debugging.

One option is to not guard at all, as in
```python
from _logging import *
from _common import *
...
```
which would give me an error, but I would not necessarily know which of the module caused it. In addition, we have the overwrite issue (see above).

The only other way I see atm is to somehow guard the imports, which I tried to be a bit more clever about and at least print some warning (see latest commit)

Reply:
How about one `module_name.py` with `from _module_name import *` per module? That can even be automated if you want to. 
BTW, isn't this a non problem if you use a debugger with stepping?

Reply:
> How about one `module_name.py` with `from _module_name import *` per module? That can even be automated if you want to.

That would, however, mean that stuff is not available in `dune.xt.common` but in `dune.xt.common.logger`, `dune.xt.common.common`, and so on.
In general, further submodules are ok with me (`dune.xt.foo` is basically one level of hierarchy and that would be the second), but we would need better names than `common`, for instance, and that would basically mean that everything is only available below an additional hierarchy.
However, this will not really work in functions, where we have one `.so` for each dim (see link above).
For instance, I would like to have the symbols from `_function_interface_1d`, `_function_interface_2d` and `_function_interface_3d` available in the same `function_interface.py` module.
So at least there we would have to use some guarded import.

Reply:
Yeah, true, more hierarchy isn't nicer. 
Bottom line for me: fewer bunched star imports would be great, import orderings need to be handled at the site (the .so) where needed, not in a different python module.

Reply:
Oops, sorry, forgot the review resolve.